sstable/reader.go

```diff
@@ -2839,56 +2892,80 @@ func (r *Reader) readBlock(
 	// interface{}, unless necessary.
 	if readDuration >= slowReadTracingThreshold && r.opts.LoggerAndTracer.IsTracingEnabled(ctx) {
 		r.opts.LoggerAndTracer.Eventf(ctx, "reading %d bytes took %s",
-			bh.Length+blockTrailerLen, readDuration.String())
+			int(bh.Length+blockTrailerLen), readDuration.String())
 	}
 	if stats != nil {
 		stats.BlockReadDuration += readDuration
 	}
 	if err != nil {
-		r.opts.Cache.Free(v)
-		return cache.Handle{}, err
+		compressed.release()
+		return bufferHandle{}, err
 	}
-
-	if err := checkChecksum(r.checksumType, b, bh, r.fileNum); err != nil {
-		r.opts.Cache.Free(v)
-		return cache.Handle{}, err
+	if err := checkChecksum(r.checksumType, compressed.get(), bh, r.fileNum); err != nil {
+		compressed.release()
+		return bufferHandle{}, err
 	}
 
-	typ := blockType(b[bh.Length])
-	b = b[:bh.Length]
-	v.Truncate(len(b))
+	typ := blockType(compressed.get()[bh.Length])
+	compressed.truncate(int(bh.Length))
 
-	decoded, err := decompressBlock(r.opts.Cache, typ, b)
-	if decoded != nil {
-		r.opts.Cache.Free(v)
-		v = decoded
-		b = v.Buf()
-	} else if err != nil {
-		r.opts.Cache.Free(v)
-		return cache.Handle{}, err
+	var decompressed cacheValueOrBuf
+	if typ == noCompressionBlockType {
+		decompressed = compressed
+	} else {
+		// Decode the length of the decompressed value.
+		decodedLen, prefixLen, err := decompressedLen(typ, compressed.get())
+		if err != nil {
+			compressed.release()
+			return bufferHandle{}, err
+		}
+
+		if bufferPool != nil {
+			decompressed = cacheValueOrBuf{buf: bufferPool.Alloc(decodedLen)}
+		} else {
+			decompressed = cacheValueOrBuf{
+				cache: r.opts.Cache,
+				v:     r.opts.Cache.Alloc(decodedLen),
+			}
+		}
+		if _, err := decompressInto(typ, compressed.get()[prefixLen:], decompressed.get()); err != nil {
+			compressed.release()
+			return bufferHandle{}, err
+		}
+		compressed.release()
 	}
 
 	if transform != nil {
-		// Transforming blocks is rare, so the extra copy of the transformed data
-		// is not problematic.
-		var err error
-		b, err = transform(b)
+		// Transforming blocks is very rare, so the extra copy of the
+		// transformed data is not problematic.
+		tmpTransformed, err := transform(decompressed.get())
```

Comment:
We were aiming to leverage transformations more for progressive restore, so ideally we'd want to clean this up in the future or have the temp block also use a pool. Nothing to do at the moment, but maybe worth a TODO if we are merging this in.